File: IsometricRPG/IsometricRPG/Items/ItemSpawner.swift
import SpriteKit

/// Responsible for placing items into newly loaded chunks based on triggers and loot tables.
final class ItemSpawner {
    private(set) var worldItems: [WorldItem] = []
    private weak var worldNode: SKNode?

    init(worldNode: SKNode) {
        self.worldNode = worldNode
    }

    // MARK: - Spawn Items for a Chunk

    /// Evaluate all triggers, build a weighted loot table, and spawn items in the chunk.
    func spawnItems(in chunk: Chunk, context: GameContext) {
        guard !chunk.spawnedItems else { return }
        chunk.spawnedItems = true

        // Collect walkable positions for item placement
        let positions: [(worldCol: Int, worldRow: Int)]
        if !chunk.rooms.isEmpty {
            positions = chunk.walkableRoomPositions()
        } else {
            positions = chunk.walkablePositions()
        }

        guard !positions.isEmpty else { return }

        // Build weighted loot table
        var weights: [ItemType: Float] = [:]
        for item in ItemType.allCases {
            weights[item] = item.baseWeight
        }

        // Apply all active triggers
        var bonusCount = 0
        for trigger in SpawnTriggerRegistry.allTriggers {
            if trigger.shouldFire(context: context) {
                for (item, multiplier) in trigger.weightModifiers {
                    weights[item, default: item.baseWeight] *= multiplier
                }
                bonusCount += trigger.spawnCountBonus
            }
        }

        // Determine spawn count
        var rng = SeededRNG(seed: chunk.coord.chunkSeed &+ 0x14E3)
        let baseCount = chunk.rooms.isEmpty ? 1 : min(chunk.rooms.count, Constants.maxItemsPerChunk)
        let totalCount = min(baseCount + bonusCount, Constants.maxItemsPerChunk)

        // Weighted random selection
        var shuffledPositions = positions
        // Fisher-Yates shuffle with seeded RNG
        for i in stride(from: shuffledPositions.count - 1, through: 1, by: -1) {
            let j = Int(rng.next() % UInt64(i + 1))
            shuffledPositions.swapAt(i, j)
        }

        for i in 0..<min(totalCount, shuffledPositions.count) {
            let itemType = weightedRandomItem(weights: weights, rng: &rng)
            let pos = shuffledPositions[i]
            let worldPos = CGPoint(x: CGFloat(pos.worldCol) + 0.5,
                                   y: CGFloat(pos.worldRow) + 0.5)
            let item = WorldItem(itemType: itemType, worldPosition: worldPos)
            worldItems.append(item)
            worldNode?.addChild(item.node)
        }
    }

    // MARK: - Item Collection

    /// Check proximity and collect items near the player.
    func collectItems(near player: Player) {
        let playerPos = player.worldPosition

        for item in worldItems where !item.isPickedUp {
            let dist = IsometricMath.distance(item.worldPosition, playerPos)

            if dist < Constants.itemPickupRange {
                item.applyEffect(to: player)
                item.pickup()
                showPickupText(item.itemType.displayName, at: item.node.position)
            } else if dist < Constants.itemAttractRange {
                // Attract toward player
                let dir = IsometricMath.direction(from: item.worldPosition, to: playerPos)
                let attractSpeed = Constants.itemAttractSpeed * (1.0 / 60.0)
                item.worldPosition = CGPoint(
                    x: item.worldPosition.x + dir.x * attractSpeed,
                    y: item.worldPosition.y + dir.y * attractSpeed
                )
                item.syncPosition()
            }
        }

        // Clean up collected items
        worldItems.removeAll { $0.isPickedUp && $0.node.parent == nil }
    }

    // MARK: - Cleanup

    /// Remove items belonging to unloaded chunks.
    func removeItems(inChunk coord: ChunkCoord) {
        let originCol = CGFloat(coord.worldOriginCol)
        let originRow = CGFloat(coord.worldOriginRow)
        let size = CGFloat(Constants.chunkSize)

        let toRemove = worldItems.filter { item in
            let x = item.worldPosition.x
            let y = item.worldPosition.y
            return x >= originCol && x < originCol + size &&
                   y >= originRow && y < originRow + size
        }

        for item in toRemove {
            item.node.removeFromParent()
        }
        worldItems.removeAll { item in
            let x = item.worldPosition.x
            let y = item.worldPosition.y
            return x >= originCol && x < originCol + size &&
                   y >= originRow && y < originRow + size
        }
    }

    func removeAll() {
        for item in worldItems {
            item.node.removeFromParent()
        }
        worldItems.removeAll()
    }

    // MARK: - Helpers

    private func weightedRandomItem(weights: [ItemType: Float], rng: inout SeededRNG) -> ItemType {
        let totalWeight = weights.values.reduce(0, +)
        var roll = Float(rng.next() % 10000) / 10000.0 * totalWeight
        for (item, weight) in weights {
            roll -= weight
            if roll <= 0 {
                return item
            }
        }
        return .healthPotion // fallback
    }

    private func showPickupText(_ text: String, at position: CGPoint) {
        guard let parent = worldNode else { return }
        let label = SKLabelNode(fontNamed: "Helvetica-Bold")
        label.text = text
        label.fontSize = 10
        label.fontColor = .white
        label.position = position
        label.zPosition = Constants.ZPosition.entity + 5
        parent.addChild(label)

        let rise = SKAction.moveBy(x: 0, y: 25, duration: 0.8)
        let fade = SKAction.fadeOut(withDuration: 0.6)
        label.run(SKAction.sequence([
            SKAction.group([rise, fade]),
            SKAction.removeFromParent()
        ]))
    }
}

File: IsometricRPGMono/Items/ItemSpawner.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;

namespace IsometricRPG;

/// <summary>
/// Responsible for placing WorldItems into newly loaded chunks and managing
/// item attraction / collection near the player.
/// </summary>
public class ItemSpawner
{
    private readonly List<WorldItem> _items = new List<WorldItem>();

    public IReadOnlyList<WorldItem> Items => _items;

    // -------------------------------------------------------------------------
    // Spawning

    /// <summary>
    /// Evaluate spawn triggers, build a weighted loot table and place up to
    /// MaxItemsPerChunk items at walkable positions inside the chunk.
    /// Does nothing if the chunk has already had items spawned.
    /// </summary>
    public void SpawnItemsForChunk(Chunk chunk, GameContext context, SeededRNG rng)
    {
        if (chunk.SpawnedItems) return;
        chunk.SpawnedItems = true;

        // Collect candidate walkable positions
        var positions = chunk.Rooms.Count > 0
            ? chunk.WalkableRoomPositions()
            : chunk.WalkablePositions();

        if (positions.Count == 0) return;

        // Build base weight table from active triggers
        var weights = SpawnTriggerRegistry.BuildWeightTable(context);

        // Fall back to equal weights for all item types if no triggers fire
        if (weights.Count == 0)
        {
            foreach (ItemType t in Enum.GetValues<ItemType>())
                weights[t] = 1f;
        }

        // Determine spawn count (1 base for open-world, room-count for dungeons)
        int baseCount = chunk.Rooms.Count == 0 ? 1 : Math.Min(chunk.Rooms.Count, Constants.MaxItemsPerChunk);
        int spawnCount = Math.Min(baseCount, Constants.MaxItemsPerChunk);

        // Fisher-Yates shuffle with the provided seeded RNG
        var shuffled = new List<(int worldCol, int worldRow)>(positions);
        for (int i = shuffled.Count - 1; i >= 1; i--)
        {
            int j = (int)(rng.Next() % (ulong)(i + 1));
            (shuffled[i], shuffled[j]) = (shuffled[j], shuffled[i]);
        }

        for (int i = 0; i < Math.Min(spawnCount, shuffled.Count); i++)
        {
            var (wCol, wRow) = shuffled[i];
            var worldPos = new Vector2(wCol + 0.5f, wRow + 0.5f);

            var itemType = WeightedRandomItem(weights, rng);
            var item = CreateItem(itemType);
            _items.Add(new WorldItem(item, worldPos));
        }
    }

    // -------------------------------------------------------------------------
    // Per-frame update

    /// <summary>
    /// Advance hover animation on all items, attract items within AttractRange,
    /// collect items within PickupRange.
    /// </summary>
    public void Update(double deltaTime, Player player)
    {
        foreach (var item in _items)
        {
            if (item.IsCollected) continue;

            item.Update(deltaTime);

            float dist = IsometricMath.Distance(item.WorldPosition, player.WorldPosition);

            if (dist < Constants.ItemPickupRange)
            {
                item.ApplyEffect(player);
            }
            else if (dist < Constants.ItemAttractRange)
            {
                var dir = IsometricMath.Direction(item.WorldPosition, player.WorldPosition);
                float attractStep = Constants.ItemAttractSpeed * (float)deltaTime;
                item.WorldPosition += dir * attractStep;
            }
        }

        // Prune collected items
        _items.RemoveAll(i => i.IsCollected);
    }

    // -------------------------------------------------------------------------
    // Chunk unload cleanup

    /// <summary>
    /// Remove all items whose world position falls inside the given chunk.
    /// </summary>
    public void RemoveItemsInChunk(ChunkCoord coord)
    {
        var (ox, oy) = coord.WorldOrigin();
        float size = Constants.ChunkSize;
        _items.RemoveAll(item =>
            item.WorldPosition.X >= ox && item.WorldPosition.X < ox + size &&
            item.WorldPosition.Y >= oy && item.WorldPosition.Y < oy + size);
    }

    public void Clear() => _items.Clear();

    // -------------------------------------------------------------------------
    // Drawing

    public void Draw(SpriteBatch spriteBatch, SpriteManager spriteManager, Vector2 cameraPos)
    {
        foreach (var item in _items)
            item.Draw(spriteBatch, spriteManager, cameraPos);
    }

    // -------------------------------------------------------------------------
    // Helpers

    private static ItemType WeightedRandomItem(Dictionary<ItemType, float> weights, SeededRNG rng)
    {
        float total = 0f;
        foreach (var w in weights.Values) total += w;

        float roll = (float)(rng.Next() % 10000UL) / 10000f * total;
        foreach (var (type, weight) in weights)
        {
            roll -= weight;
            if (roll <= 0f) return type;
        }

        return ItemType.HealthPotion; // fallback
    }

    private static Item CreateItem(ItemType type)
    {
        return new Item(
            id:          type.ToString().ToLowerInvariant(),
            name:        ItemTypeName(type),
            description: string.Empty,
            type:        type,
            rarity:      Rarity.Common,
            value:       10,
            weight:      1f);
    }

    private static string ItemTypeName(ItemType type)
    {
        return type switch
        {
            ItemType.HealthPotion => "Health Potion",
            ItemType.ManaPotion   => "Mana Potion",
            ItemType.Antidote     => "Antidote",
            ItemType.Food         => "Food",
            ItemType.Sword        => "Sword",
            ItemType.Bow          => "Bow",
            ItemType.Staff        => "Staff",
            ItemType.Dagger       => "Dagger",
            ItemType.Axe          => "Axe",
            ItemType.Helmet       => "Helmet",
            ItemType.ChestArmor   => "Chest Armor",
            ItemType.Leggings     => "Leggings",
            ItemType.Boots        => "Boots",
            ItemType.Shield       => "Shield",
            ItemType.Gold         => "Gold",
            ItemType.Gem          => "Gem",
            ItemType.Artifact     => "Artifact",
            ItemType.QuestScroll  => "Quest Scroll",
            ItemType.AncientKey   => "Ancient Key",
            ItemType.CrystalShard => "Crystal Shard",
            ItemType.Wood         => "Wood",
            ItemType.Stone        => "Stone",
            ItemType.Iron         => "Iron",
            ItemType.Leather      => "Leather",
            ItemType.Cloth        => "Cloth",
            _                     => type.ToString()
        };
    }
}
